### PR TITLE
Add newfeatures and updateto es6

### PR DESCRIPTION
**Pull request by jomi-se**
* Adds new method `addAndFilterDescendants(node, descendants, filterFunc)` to tarjan-graph
* Updates code style to ES6 and changes indentation do 2-spaces
* Removed some unneeded code (appending of empty array to descendants and reversal of descendants)
* Added and updated unit-tests (all green)

### CODE REVIEW

**Comment by jomi-se** [comment]
Care to take a quick look before I merge into master ? @ebourmalo @ArthurMialon 

The diff is pretty ugly because I changed the indentation style 🙈

File: tests/graph-tests.js [hunk 1]
```diff
@@ -14,9 +14,9 @@ describe('Graph', function() {
 		cycles.should.have.length(1);
 		graph.hasCycle().should.equal(true);
 		var cycle = cycles[0];
-		cycle[0].should.have.property('name', 'd');
-		cycle[1].should.have.property('name', 'c');
-		cycle[2].should.have.property('name', 'e');
+		cycle[0].should.have.property('name', 'c');
+		cycle[1].should.have.property('name', 'e');
+		cycle[2].should.have.property('name', 'd');
```
Comment (ebourmalo) [inline tests/graph-tests.js]: you changed the node identifiers ?? 
`d, c, e` VS `c, e, d`

File: tests/graph-tests.js [hunk 2]
```diff
@@ -100,24 +127,24 @@ describe('Graph', function() {
...
   g -> h\n\
-  f -> g\n\
-  f -> a\n\
+  g -> i\n\
   f -> c\n\
+  f -> a\n\
+  f -> g\n\
   h -> j\n\
   i -> j\n\
   j -> f\n\
```
Comment (ebourmalo) [inline tests/graph-tests.js]: you should update the indent for this file too :)

File: index.js [hunk 3]
```diff
@@ -1,198 +1,212 @@
...
-		});
-
-		lines.push('}');
-		return lines.join('\n') + '\n';
-	}
-};
-
-module.exports = Graph;
+  add(key, descendants) {
+    descendants = Array.isArray(descendants) ? descendants : [descendants]
```
Comment (ArthurMialon) [inline index.js]: @jomi-se Don't know about perfs but this snippet could be
```
descendants = [].concat(descendants)

File: index.js [hunk 4]
```diff
@@ -1,198 +1,212 @@
...
+
+    const successors = descendants.map((descendant) => {
```
Comment (ArthurMialon) [inline index.js]: @jomi-se You can remove brackets here when you have 1 argument only

File: index.js [hunk 5]
```diff
@@ -1,198 +1,212 @@
...
+    if (!this.vertices[key]) {
+      this.vertices[key] = new Vertex(key)
+    }
+
+    this.vertices[key].successors = successors
+    return this
+  },
+
+  addAndFilterDescendants(key, descendants, filter) {
+    descendants = Array.isArray(descendants) ? descendants : [descendants]
```
Comment (ArthurMialon) [inline index.js]: @jomi-se Same here

File: index.js [hunk 6]
```diff
@@ -1,198 +1,212 @@
...
+    if (!this.vertices[key]) {
+      this.vertices[key] = new Vertex(key)
+    }
+
+    this.vertices[key].successors = successors
+    return this
+  },
+
+  reset() {
+    Object.keys(this.vertices).forEach((key) => {
```
Comment (ArthurMialon) [inline index.js]: @jomi-se Remove bracket

File: index.js [hunk 7]
```diff
@@ -1,198 +1,212 @@
...
+  },
+
+  getCycles() {
+    return this.getStronglyConnectedComponents().filter((scc) => scc.length > 1)
+  },
+
+  clone() {
+    const graph = new Graph()
+
+    Object.keys(this.vertices).forEach((key) => {
```
Comment (ArthurMialon) [inline index.js]: @jomi-se remove brackets :)

File: index.js [hunk 8]
```diff
@@ -1,198 +1,212 @@
...
+
+    const cycles = this.getCycles()
+    cycles.forEach((scc, i) => {
+      lines.push('  subgraph cluster' + i + ' {')
+      lines.push('    color=red;')
+      lines.push('    ' + scc.map((v) => v.name).join('; ') + ';')
+      lines.push('  }')
+    })
+
+    Object.keys(V).forEach((key) => {
```
Comment (ArthurMialon) [inline index.js]: @jomi-se removable brackets

File: tests/graph-tests.js [hunk 9]
```diff
@@ -47,9 +47,36 @@ describe('Graph', function() {
...
-		graph.getDescendants('c').should.eql([ 'b', 'd', 'e' ]);
+		graph.getDescendants('a').should.eql([ 'c', 'b', 'e', 'd' ]);
+		graph.getDescendants('b').should.eql([ 'e', 'd' ]);
+		graph.getDescendants('c').should.eql([ 'b', 'e', 'd' ]);
+		graph.getDescendants('d').should.eql([ 'e' ]);
+		graph.getDescendants('e').should.eql([ ]);
+	});
+
+	it('should filter descendants correctly', function() {
+	 	var filterFunc = (desc) => {
```
Comment (ArthurMialon) [inline tests/graph-tests.js]: @jomi-se Remove brackets

File: tests/graph-tests.js (hunk 1 above)
Comment (jomi-se) [inline tests/graph-tests.js]: Yes, I removed some code that reversed the order of descendants for every single node.

There is no functional impact to this change, but the unit tests that checked for descendants needed to be updated to this change

**Comment by ebourmalo** [comment]
just add a file `.eslintrc` at the root level with the following content: 
```js
{
  "extends": "payfit"
}
```

**Comment by jomi-se** [comment]
No changes needed for the `package.json` file?

**Comment by jomi-se** [comment]
Nice catches, merci. Done @ebourmalo @ArthurMialon

**Comment by ebourmalo** [comment]
You need to add all the `eslint-xx` deps, you'll find them in the repo `apps` in the same file

File: tests/graph-tests.js [hunk 10]
```diff
@@ -1,161 +1,187 @@
...
-
-		var expected = 'digraph {\n\
+}\n'
+
+    graph.toDot().should.equal(expected)
+  })
+
+  it('should convert cyclic graph to DOT showing SCCs', function() {
+    const graph = new Graph()
+        .add('a', ['b', 'c'])
```
Comment (ebourmalo) [inline tests/graph-tests.js]: all the indent with `.add` are wrong @jomi-se ^^

Comment (jomi-se) [inline tests/graph-tests.js]: It should be only 2 extra spaces for indentation right ? Strange, eslint is not showing it as an error -_-

Comment (jomi-se) [inline tests/graph-tests.js]: Yeah, I get the same result on the `apps` repo. If I follow with a new line after a `new Object()` declaration my editor adds extra spaces but eslint doesn't complain for some reason.

**Pull request merged**